# Automatización de análisis de consultas Fosec (Conacyt)

> Por: Jesús Octavio García Flores,  octavio5545@ciencias.unam.mx,  2022

In [1]:
import pandas as pd
import numpy as np

In [2]:
url = "Datase_general_fosec.csv"

In [3]:
datos = pd.read_csv(url)
df = pd.DataFrame(datos)

In [4]:
df

CLAVE_FONDO DESCRIPCION_CORTA_FONDO  \
0          FSCNS01         FOSEC SEGOB-CNS   
1          FSCNS01         FOSEC SEGOB-CNS   
2          FSCNS01         FOSEC SEGOB-CNS   
3          FSCNS01         FOSEC SEGOB-CNS   
4          FSCNS01         FOSEC SEGOB-CNS   
...            ...                     ...   
392394     FSCNS01         FOSEC SEGOB-CNS   
392395     FSCNS01         FOSEC SEGOB-CNS   
392396     FSCNS01         FOSEC SEGOB-CNS   
392397     FSCNS01         FOSEC SEGOB-CNS   
392398     FSCNS01         FOSEC SEGOB-CNS   

                                  DESCRIPCION_LARGA_FONDO  \
0       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
1       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
2       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
3       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
4       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
...                                                   ...   
392394  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392395  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392396  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392397  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392398  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   

               FECHA_ALTA_FONDO  CVE_CONVOCATORIA  \
0       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
1       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
2       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
3       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
4       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
...                         ...               ...   
392394  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392395  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392396  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392397  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392398  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   

       DESCRIPCION_LARGA_CONVOCATORIA FECHA_INICIO_CONVOCATORIA  \
0                 CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
1                 CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
2                 CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
3                 CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
4                 CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
...                               ...                       ...   
392394            CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
392395            CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
392396            CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
392397            CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
392398            CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   

         FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
0                           NaN               B   
1       2019-02-08 23:59:59.000               B   
2       2019-01-18 18:00:00.000               B   
3       2019-05-24 23:59:59.000               B   
4       2019-05-03 18:00:00.000               B   
...                         ...             ...   
392394  2019-02-08 23:59:59.000               B   
392395  2019-01-18 18:00:00.000               B   
392396  2019-05-24 23:59:59.000               B   
392397  2019-05-03 18:00:00.000               B   
392398                      NaN               B   

                      DESCRIPCION_MODALIDAD CLAVE_PROYECTO  
0       Desarrollo tecnológico e innovación       B-S-7781  
1       Desarrollo tecnológico e innovación       B-S-7781  
2       Desarrollo tecnológico e innovación       B-S-7781  
3       Desarrollo tecnológico e innovación       B-S-7781  
4       Desarrollo tecnológico e innovación       B-S-7781  
...                                     ...            ...  
392394  Desarrollo tecnológico e innovación     B-S-132604  
392395  Desarrollo tecnológico e innovación     B-S-132604  
392396  Desarrollo tecnológico e innovación     B-S-132604  
392397  Desarrollo tecnológico e innovación     B-S-1326

In [5]:
df.shape

(392399, 11)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 392399 entries, 0 to 392398
Data columns (total 11 columns):
 #   Column                          Non-Null Count   Dtype 
---  ------                          --------------   ----- 
 0   CLAVE_FONDO                     392399 non-null  object
 1   DESCRIPCION_CORTA_FONDO         392399 non-null  object
 2   DESCRIPCION_LARGA_FONDO         392399 non-null  object
 3   FECHA_ALTA_FONDO                392399 non-null  object
 4   CVE_CONVOCATORIA                392399 non-null  object
 5   DESCRIPCION_LARGA_CONVOCATORIA  392399 non-null  object
 6   FECHA_INICIO_CONVOCATORIA       392399 non-null  object
 7   FECHA_FIN_CONVOCATORIA          274465 non-null  object
 8   CLAVE_MODALIDAD                 392399 non-null  object
 9   DESCRIPCION_MODALIDAD           392399 non-null  object
 10  CLAVE_PROYECTO                  392399 non-null  object
dtypes: object(11)
memory usage: 32.9+ MB


In [7]:
df.nunique()

CLAVE_FONDO                          6
DESCRIPCION_CORTA_FONDO              6
DESCRIPCION_LARGA_FONDO              6
FECHA_ALTA_FONDO                     4
CVE_CONVOCATORIA                     8
DESCRIPCION_LARGA_CONVOCATORIA       4
FECHA_INICIO_CONVOCATORIA           32
FECHA_FIN_CONVOCATORIA              26
CLAVE_MODALIDAD                      2
DESCRIPCION_MODALIDAD                2
CLAVE_PROYECTO                    1610
dtype: int64

In [8]:
df.isnull().sum()

CLAVE_FONDO                            0
DESCRIPCION_CORTA_FONDO                0
DESCRIPCION_LARGA_FONDO                0
FECHA_ALTA_FONDO                       0
CVE_CONVOCATORIA                       0
DESCRIPCION_LARGA_CONVOCATORIA         0
FECHA_INICIO_CONVOCATORIA              0
FECHA_FIN_CONVOCATORIA            117934
CLAVE_MODALIDAD                        0
DESCRIPCION_MODALIDAD                  0
CLAVE_PROYECTO                         0
dtype: int64

#### Son 117934 registros sin fecha de finalización de convocatoria

In [9]:
df_ = df.drop_duplicates()
dif = df.shape[0] - df_.shape[0]

print("\nDimensiones del DataFrame original: \t\t\t {} \n".format(df.shape))
print("\nDimensiones del DataFrame con duplicados eliminados:  {} \n".format(df_.shape))
if dif !=0:
    print("\nSe eliminaron {}  filas del DataFrame original.".format(dif) )
else:
    print("\nEste Dataframe NO contiene filas duplicadas.")



Dimensiones del DataFrame original: 			 (392399, 11) 


Dimensiones del DataFrame con duplicados eliminados:  (10524, 11) 


Se eliminaron 381875  filas del DataFrame original.


In [10]:
df_.to_csv("Datase_general_FOSEC_(reducido).csv")

#### Dataframe reducido:

In [11]:
df_

CLAVE_FONDO DESCRIPCION_CORTA_FONDO  \
0          FSCNS01         FOSEC SEGOB-CNS   
1          FSCNS01         FOSEC SEGOB-CNS   
2          FSCNS01         FOSEC SEGOB-CNS   
3          FSCNS01         FOSEC SEGOB-CNS   
4          FSCNS01         FOSEC SEGOB-CNS   
...            ...                     ...   
392388     FSCNS01         FOSEC SEGOB-CNS   
392389     FSCNS01         FOSEC SEGOB-CNS   
392390     FSCNS01         FOSEC SEGOB-CNS   
392391     FSCNS01         FOSEC SEGOB-CNS   
392392     FSCNS01         FOSEC SEGOB-CNS   

                                  DESCRIPCION_LARGA_FONDO  \
0       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
1       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
2       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
3       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
4       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
...                                                   ...   
392388  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392389  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392390  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392391  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392392  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   

               FECHA_ALTA_FONDO  CVE_CONVOCATORIA  \
0       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
1       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
2       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
3       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
4       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
...                         ...               ...   
392388  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392389  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392390  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392391  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392392  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   

       DESCRIPCION_LARGA_CONVOCATORIA FECHA_INICIO_CONVOCATORIA  \
0                 CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
1                 CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
2                 CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
3                 CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
4                 CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
...                               ...                       ...   
392388            CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
392389            CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
392390            CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
392391            CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
392392            CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   

         FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
0                           NaN               B   
1       2019-02-08 23:59:59.000               B   
2       2019-01-18 18:00:00.000               B   
3       2019-05-24 23:59:59.000               B   
4       2019-05-03 18:00:00.000               B   
...                         ...             ...   
392388  2019-02-08 23:59:59.000               B   
392389  2019-01-18 18:00:00.000               B   
392390  2019-05-24 23:59:59.000               B   
392391  2019-05-03 18:00:00.000               B   
392392                      NaN               B   

                      DESCRIPCION_MODALIDAD CLAVE_PROYECTO  
0       Desarrollo tecnológico e innovación       B-S-7781  
1       Desarrollo tecnológico e innovación       B-S-7781  
2       Desarrollo tecnológico e innovación       B-S-7781  
3       Desarrollo tecnológico e innovación       B-S-7781  
4       Desarrollo tecnológico e innovación       B-S-7781  
...                                     ...            ...  
392388  Desarrollo tecnológico e innovación     B-S-132604  
392389  Desarrollo tecnológico e innovación     B-S-132604  
392390  Desarrollo tecnológico e innovación     B-S-132604  
392391  Desarrollo tecnológico e innovación     B-S-1326

In [12]:
df_.nunique()

CLAVE_FONDO                          6
DESCRIPCION_CORTA_FONDO              6
DESCRIPCION_LARGA_FONDO              6
FECHA_ALTA_FONDO                     4
CVE_CONVOCATORIA                     8
DESCRIPCION_LARGA_CONVOCATORIA       4
FECHA_INICIO_CONVOCATORIA           32
FECHA_FIN_CONVOCATORIA              26
CLAVE_MODALIDAD                      2
DESCRIPCION_MODALIDAD                2
CLAVE_PROYECTO                    1610
dtype: int64

In [13]:
df_.isnull().sum()

CLAVE_FONDO                          0
DESCRIPCION_CORTA_FONDO              0
DESCRIPCION_LARGA_FONDO              0
FECHA_ALTA_FONDO                     0
CVE_CONVOCATORIA                     0
DESCRIPCION_LARGA_CONVOCATORIA       0
FECHA_INICIO_CONVOCATORIA            0
FECHA_FIN_CONVOCATORIA            3220
CLAVE_MODALIDAD                      0
DESCRIPCION_MODALIDAD                0
CLAVE_PROYECTO                       0
dtype: int64

#### Según los datos del dataframe reducido, existen 3220 convocatorias que no tienen fecha de finalización.

### Filas repetidas:


In [14]:
df[df.duplicated()]

CLAVE_FONDO DESCRIPCION_CORTA_FONDO  \
6          FSCNS01         FOSEC SEGOB-CNS   
7          FSCNS01         FOSEC SEGOB-CNS   
8          FSCNS01         FOSEC SEGOB-CNS   
9          FSCNS01         FOSEC SEGOB-CNS   
10         FSCNS01         FOSEC SEGOB-CNS   
...            ...                     ...   
392394     FSCNS01         FOSEC SEGOB-CNS   
392395     FSCNS01         FOSEC SEGOB-CNS   
392396     FSCNS01         FOSEC SEGOB-CNS   
392397     FSCNS01         FOSEC SEGOB-CNS   
392398     FSCNS01         FOSEC SEGOB-CNS   

                                  DESCRIPCION_LARGA_FONDO  \
6       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
7       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
8       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
9       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
10      Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
...                                                   ...   
392394  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392395  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392396  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392397  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392398  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   

               FECHA_ALTA_FONDO  CVE_CONVOCATORIA  \
6       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
7       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
8       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
9       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
10      2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
...                         ...               ...   
392394  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392395  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392396  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392397  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392398  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   

       DESCRIPCION_LARGA_CONVOCATORIA FECHA_INICIO_CONVOCATORIA  \
6                 CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
7                 CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
8                 CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
9                 CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
10                CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
...                               ...                       ...   
392394            CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
392395            CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
392396            CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
392397            CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
392398            CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   

         FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
6                           NaN               B   
7       2019-02-08 23:59:59.000               B   
8       2019-01-18 18:00:00.000               B   
9       2019-05-24 23:59:59.000               B   
10      2019-05-03 18:00:00.000               B   
...                         ...             ...   
392394  2019-02-08 23:59:59.000               B   
392395  2019-01-18 18:00:00.000               B   
392396  2019-05-24 23:59:59.000               B   
392397  2019-05-03 18:00:00.000               B   
392398                      NaN               B   

                      DESCRIPCION_MODALIDAD CLAVE_PROYECTO  
6       Desarrollo tecnológico e innovación       B-S-7781  
7       Desarrollo tecnológico e innovación       B-S-7781  
8       Desarrollo tecnológico e innovación       B-S-7781  
9       Desarrollo tecnológico e innovación       B-S-7781  
10      Desarrollo tecnológico e innovación       B-S-7781  
...                                     ...            ...  
392394  Desarrollo tecnológico e innovación     B-S-132604  
392395  Desarrollo tecnológico e innovación     B-S-132604  
392396  Desarrollo tecnológico e innovación     B-S-132604  
392397  Desarrollo tecnológico e innovación     B-S-1326

### Convocatorias que no tienen fecha de finalización:

In [15]:
df[df.FECHA_FIN_CONVOCATORIA.isnull()]

CLAVE_FONDO DESCRIPCION_CORTA_FONDO  \
0          FSCNS01         FOSEC SEGOB-CNS   
5          FSCNS01         FOSEC SEGOB-CNS   
6          FSCNS01         FOSEC SEGOB-CNS   
11         FSCNS01         FOSEC SEGOB-CNS   
12         FSCNS01         FOSEC SEGOB-CNS   
...            ...                     ...   
392386     FSCNS01         FOSEC SEGOB-CNS   
392387     FSCNS01         FOSEC SEGOB-CNS   
392392     FSCNS01         FOSEC SEGOB-CNS   
392393     FSCNS01         FOSEC SEGOB-CNS   
392398     FSCNS01         FOSEC SEGOB-CNS   

                                  DESCRIPCION_LARGA_FONDO  \
0       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
5       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
6       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
11      Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
12      Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
...                                                   ...   
392386  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392387  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392392  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392393  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392398  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   

               FECHA_ALTA_FONDO  CVE_CONVOCATORIA  \
0       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
5       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
6       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
11      2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
12      2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
...                         ...               ...   
392386  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392387  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392392  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392393  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392398  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   

       DESCRIPCION_LARGA_CONVOCATORIA FECHA_INICIO_CONVOCATORIA  \
0                 CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
5                 CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   
6                 CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
11                CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   
12                CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
...                               ...                       ...   
392386            CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   
392387            CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
392392            CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   
392393            CONVOCATORIA-2017-6   2019-05-27 00:00:00.000   
392398            CONVOCATORIA-2017-6   2017-12-18 08:00:00.000   

       FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
0                         NaN               B   
5                         NaN               B   
6                         NaN               B   
11                        NaN               B   
12                        NaN               B   
...                       ...             ...   
392386                    NaN               B   
392387                    NaN               B   
392392                    NaN               B   
392393                    NaN               B   
392398                    NaN               B   

                      DESCRIPCION_MODALIDAD CLAVE_PROYECTO  
0       Desarrollo tecnológico e innovación       B-S-7781  
5       Desarrollo tecnológico e innovación       B-S-7781  
6       Desarrollo tecnológico e innovación       B-S-7781  
11      Desarrollo tecnológico e innovación       B-S-7781  
12      Desarrollo tecnológico e innovación       B-S-8064  
...                                     ...            ...  
392386  Desarrollo tecnológico e innovación     B-S-132304  
392387  Desarrollo tecnológico e innovación     B-S-132604  
392392  Desarrollo tecnológico e innovación     B-S-132604  
392393  Desarrollo tecnológico e innovación     B-S-132604  
392398  Desarrollo 

### Convocatorias que sí tienen fecha de finalización:

In [16]:
df[~df.FECHA_FIN_CONVOCATORIA.isnull()]

CLAVE_FONDO DESCRIPCION_CORTA_FONDO  \
1          FSCNS01         FOSEC SEGOB-CNS   
2          FSCNS01         FOSEC SEGOB-CNS   
3          FSCNS01         FOSEC SEGOB-CNS   
4          FSCNS01         FOSEC SEGOB-CNS   
7          FSCNS01         FOSEC SEGOB-CNS   
...            ...                     ...   
392391     FSCNS01         FOSEC SEGOB-CNS   
392394     FSCNS01         FOSEC SEGOB-CNS   
392395     FSCNS01         FOSEC SEGOB-CNS   
392396     FSCNS01         FOSEC SEGOB-CNS   
392397     FSCNS01         FOSEC SEGOB-CNS   

                                  DESCRIPCION_LARGA_FONDO  \
1       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
2       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
3       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
4       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
7       Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
...                                                   ...   
392391  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392394  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392395  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392396  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   
392397  Fondo Sectorial CONACYT-SEGOB-CNS para la Segu...   

               FECHA_ALTA_FONDO  CVE_CONVOCATORIA  \
1       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
2       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
3       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
4       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
7       2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
...                         ...               ...   
392391  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392394  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392395  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392396  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   
392397  2017-06-16 01:49:54.000  FSCNS01-C-2017-6   

       DESCRIPCION_LARGA_CONVOCATORIA FECHA_INICIO_CONVOCATORIA  \
1                 CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
2                 CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
3                 CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
4                 CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
7                 CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
...                               ...                       ...   
392391            CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   
392394            CONVOCATORIA-2017-6   2019-01-21 00:00:00.000   
392395            CONVOCATORIA-2017-6   2018-11-12 08:00:00.000   
392396            CONVOCATORIA-2017-6   2019-05-06 00:00:00.000   
392397            CONVOCATORIA-2017-6   2019-02-18 08:00:00.000   

         FECHA_FIN_CONVOCATORIA CLAVE_MODALIDAD  \
1       2019-02-08 23:59:59.000               B   
2       2019-01-18 18:00:00.000               B   
3       2019-05-24 23:59:59.000               B   
4       2019-05-03 18:00:00.000               B   
7       2019-02-08 23:59:59.000               B   
...                         ...             ...   
392391  2019-05-03 18:00:00.000               B   
392394  2019-02-08 23:59:59.000               B   
392395  2019-01-18 18:00:00.000               B   
392396  2019-05-24 23:59:59.000               B   
392397  2019-05-03 18:00:00.000               B   

                      DESCRIPCION_MODALIDAD CLAVE_PROYECTO  
1       Desarrollo tecnológico e innovación       B-S-7781  
2       Desarrollo tecnológico e innovación       B-S-7781  
3       Desarrollo tecnológico e innovación       B-S-7781  
4       Desarrollo tecnológico e innovación       B-S-7781  
7       Desarrollo tecnológico e innovación       B-S-7781  
...                                     ...            ...  
392391  Desarrollo tecnológico e innovación     B-S-132604  
392394  Desarrollo tecnológico e innovación     B-S-132604  
392395  Desarrollo tecnológico e innovación     B-S-132604  
392396  Desarrollo tecnológico e innovación     B-S-1326

## Las descripciones (corta y larga) parecen mantener relación uno a uno con los fondos.

In [17]:
fondo_desc_corta = df[['CLAVE_FONDO', 'DESCRIPCION_CORTA_FONDO']]
fondo_desc_corta.groupby(['CLAVE_FONDO']).nunique()

DESCRIPCION_CORTA_FONDO
CLAVE_FONDO                           
FSAEM01                              1
FSCNS01                              1
FSCONAGUA01                          1
FSINMUJERES01                        1
FSSEMARNAT01                         1
FSSS01                               1

In [18]:
fondo_desc_corta.groupby(['DESCRIPCION_CORTA_FONDO']).nunique()

CLAVE_FONDO
DESCRIPCION_CORTA_FONDO             
FOSEC AEM                          1
FOSEC CONAGUA                      1
FOSEC INMUJERES                    1
FOSEC SEGOB-CNS                    1
FOSEC SEMARNAT                     1
FOSEC SS/IMSS/ISSSTE               1

In [19]:
fondo_desc_larga = df[['CLAVE_FONDO', 'DESCRIPCION_LARGA_FONDO']]
fondo_desc_larga.groupby(['CLAVE_FONDO']).nunique()

DESCRIPCION_LARGA_FONDO
CLAVE_FONDO                           
FSAEM01                              1
FSCNS01                              1
FSCONAGUA01                          1
FSINMUJERES01                        1
FSSEMARNAT01                         1
FSSS01                               1

In [20]:
fondo_desc_larga.groupby(['DESCRIPCION_LARGA_FONDO']).nunique()

CLAVE_FONDO
DESCRIPCION_LARGA_FONDO                                        
Fondo Sectorial CONACYT-SEGOB-CNS para la Segur...            1
Fondo Sectorial de Investigación Ambiental SEMA...            1
Fondo Sectorial de Investigación en Salud y Seg...            1
Fondo Sectorial de Investigación y Desarrollo s...            1
Fondo Sectorial de Investigación, Desarrollo Te...            1
MIIC.FONDO SECTORIAL INMUJERES                                1

.

* Se comprueba la unicidad de entre las claves fondo y sus descripciones.

.

In [21]:
fondo_conv = df[['CLAVE_FONDO', 'CVE_CONVOCATORIA']].groupby(['CLAVE_FONDO']).apply(lambda f1: print(f1) )

       CLAVE_FONDO  CVE_CONVOCATORIA
340533     FSAEM01  FSAEM01-C-2018-1
340534     FSAEM01  FSAEM01-C-2018-1
340535     FSAEM01  FSAEM01-C-2018-1
340536     FSAEM01  FSAEM01-C-2018-1
340537     FSAEM01  FSAEM01-C-2018-1
...            ...               ...
374178     FSAEM01  FSAEM01-C-2018-1
374179     FSAEM01  FSAEM01-C-2018-1
374180     FSAEM01  FSAEM01-C-2018-1
374181     FSAEM01  FSAEM01-C-2018-1
374182     FSAEM01  FSAEM01-C-2018-1

[1932 rows x 2 columns]
       CLAVE_FONDO  CVE_CONVOCATORIA
0          FSCNS01  FSCNS01-C-2017-6
1          FSCNS01  FSCNS01-C-2017-6
2          FSCNS01  FSCNS01-C-2017-6
3          FSCNS01  FSCNS01-C-2017-6
4          FSCNS01  FSCNS01-C-2017-6
...            ...               ...
392394     FSCNS01  FSCNS01-C-2017-6
392395     FSCNS01  FSCNS01-C-2017-6
392396     FSCNS01  FSCNS01-C-2017-6
392397     FSCNS01  FSCNS01-C-2017-6
392398     FSCNS01  FSCNS01-C-2017-6

[408 rows x 2 columns]
        CLAVE_FONDO      CVE_CONVOCATORIA
340413  FSCONAGUA01  

In [22]:
#OBTIENE LOS SUBGRUPOS DE CONVOCATORIA POR CADA FONDO

groups = df[['CLAVE_FONDO', 'CVE_CONVOCATORIA']].groupby(['CLAVE_FONDO','CVE_CONVOCATORIA'])
for group_key, group_value in groups:
    print("Group_key:", group_key, "Group_value:", group_value)
    group = groups.get_group(group_key)

    print("")


Group_key: ('FSAEM01', 'FSAEM01-C-2018-1') Group_value:        CLAVE_FONDO  CVE_CONVOCATORIA
340533     FSAEM01  FSAEM01-C-2018-1
340534     FSAEM01  FSAEM01-C-2018-1
340535     FSAEM01  FSAEM01-C-2018-1
340536     FSAEM01  FSAEM01-C-2018-1
340537     FSAEM01  FSAEM01-C-2018-1
...            ...               ...
374178     FSAEM01  FSAEM01-C-2018-1
374179     FSAEM01  FSAEM01-C-2018-1
374180     FSAEM01  FSAEM01-C-2018-1
374181     FSAEM01  FSAEM01-C-2018-1
374182     FSAEM01  FSAEM01-C-2018-1

[1932 rows x 2 columns]

Group_key: ('FSCNS01', 'FSCNS01-C-2017-6') Group_value:        CLAVE_FONDO  CVE_CONVOCATORIA
0          FSCNS01  FSCNS01-C-2017-6
1          FSCNS01  FSCNS01-C-2017-6
2          FSCNS01  FSCNS01-C-2017-6
3          FSCNS01  FSCNS01-C-2017-6
4          FSCNS01  FSCNS01-C-2017-6
...            ...               ...
392394     FSCNS01  FSCNS01-C-2017-6
392395     FSCNS01  FSCNS01-C-2017-6
392396     FSCNS01  FSCNS01-C-2017-6
392397     FSCNS01  FSCNS01-C-2017-6
392398     

In [23]:
#DEVUELVE DATAFRAME CON MÍNIMOS Y MÁXIMOS DE UN CAMPO.

def min_max_subgroups(df, cols, key, names_groups=None, save=False):
    #import pandas as pd
    
    # cols: Son las columnas del df que se quieren analizar
    # key : Nombre de la columna de agrupación
    # names_groups: lista de dos entradas con los nombres de los dataframes
    #               con los mínimos y máximos respectivamente.
    # save: True si se quiere generar los csv con mínimos y máximos.
    
    grouped_df = df[cols].groupby(cols)\
    .size().reset_index(name="Count")
    
    min_group = grouped_df.groupby([key]).min()
    max_group = grouped_df.groupby([key]).max()
    if save:
        min_group.to_csv(names_groups[0])
        max_group.to_csv(names_groups[1])
    
    return min_group, max_group


In [24]:
fond_conv_min, fond_conv_max = min_max_subgroups(df, cols=['CLAVE_FONDO', 'CVE_CONVOCATORIA'], key='CLAVE_FONDO', 
                                                     names_groups=["Min_conv_por_fondo_FOMIX.csv", "Max_conv_por_fondo_FOMIX.csv"],
                                                     save=True)


### Convocatorias que aparecen menos veces por fondo:

In [25]:
fond_conv_min

CVE_CONVOCATORIA  Count
CLAVE_FONDO                                 
FSAEM01              FSAEM01-C-2018-1   1932
FSCNS01              FSCNS01-C-2017-6    408
FSCONAGUA01      FSCONAGUA01-C-2018-1   3900
FSINMUJERES01  FSINMUJERES01-C-2018-1  11040
FSSEMARNAT01    FSSEMARNAT01-C-2018-1   5760
FSSS01                FSSS01-C-2018-1   1127

### Convocatorias que aparecen más veces por fondo:

In [26]:
fond_conv_max

CVE_CONVOCATORIA   Count
CLAVE_FONDO                                  
FSAEM01              FSAEM01-C-2018-1    1932
FSCNS01              FSCNS01-C-2017-6     408
FSCONAGUA01      FSCONAGUA01-C-2018-1    3900
FSINMUJERES01  FSINMUJERES01-C-2018-1   11040
FSSEMARNAT01    FSSEMARNAT01-C-2018-1    5760
FSSS01                FSSS01-C-2018-3  267120

### Número máximo y mínimo de convocatorias en los fondos:

In [27]:
fondos_y_conv = df[['CLAVE_FONDO', 'CVE_CONVOCATORIA']]
conv_unicas_por_fondo = fondos_y_conv.groupby('CLAVE_FONDO').nunique()
max_conv = int(conv_unicas_por_fondo.max())
min_conv = int(conv_unicas_por_fondo.min())

In [28]:
#FONDOS QUE TIENEN EL MÍNIMO DE CONVOCATORIAS ÚNICAS:
conv_unicas_por_fondo[conv_unicas_por_fondo['CVE_CONVOCATORIA'] == min_conv]

CVE_CONVOCATORIA
CLAVE_FONDO                    
FSAEM01                       1
FSCNS01                       1
FSCONAGUA01                   1
FSINMUJERES01                 1
FSSEMARNAT01                  1

In [29]:
#FONDOS QUE TIENEN EL MÁXIMO DE CONVOCATORIAS ÚNICAS:
conv_unicas_por_fondo[conv_unicas_por_fondo['CVE_CONVOCATORIA'] == max_conv]

CVE_CONVOCATORIA
CLAVE_FONDO                  
FSSS01                      3